content/6_Advanced/String_Search.mdx

```diff
@@ -175,28 +175,67 @@ We would then have these two cases:
 #include <bits/stdc++.h>
 using namespace std;
 
-int main() {
-	string s;
-	cin >> s;
+vector<int> z_function(const string &s) {
 	vector<int> z(s.size());
+	z[0] = s.size();
 
-	for (int i = 1, l = 0, r = 0; i < s.size(); i++) {
+	int l = 0, r = 0;
+	for (int i = 1; i < s.size(); i++) {
 		z[i] = max(0, min(z[i - l], r - i + 1));
 
 		while (i + z[i] < s.size() && s[z[i]] == s[i + z[i]]) {
 			l = i;
 			r = i + z[i];
 			z[i]++;
 		}
+	}
+
+	return z;
+}
 
-		if (z[i] + i == s.size()) { cout << i << ' '; }
+int main() {
+	string s;
+	cin >> s;
+	vector<int> z = z_function(s);
+
+	for (int i = 1; i < s.size(); i++) {
+		if (z[i] + i == s.size()) { cout << i << " "; }
 	}
 
 	cout << s.size() << endl;
 }
 ```
 
 </CPPSection>
+<PySection>
+
+```py
+from typing import List
+
+
+def z_function(s: str) -> List[int]:
+	n = len(s)
+	z = [n]
+	l, r = 0, 0
+	for i in range(1, n):
+		zi = min(r - i, z[i - l]) if i < r else 0
+		while i + zi < n and s[i + zi] == s[zi]:
+			zi += 1
+		if i + zi > r:
+			l, r = i, i + zi
```

Comment:
noticing a small diff in impl between this while loop & the one in the c++ version